### PR TITLE
Add gemm solver in MIOpen

### PR DESCRIPTION
**Pull request by bghimireamd**

### CODE REVIEW

File: driver/tensor_driver.hpp [hunk 1]
```diff
@@ -26,7 +26,8 @@
 #ifndef GUARD_MIOPEN_TENSOR_DRIVER_HPP
 #define GUARD_MIOPEN_TENSOR_DRIVER_HPP
 
-#include "driver.hpp"
+//#include "driver.hpp"
+#define UNPACK_VEC4(v) (v[0]), (v[1]), (v[2]), (v[3])
```
Comment (bghimireamd) [inline driver/tensor_driver.hpp:30]: You can ignore this for now. I have create another PR #1889 that address this issue.

File: test/random.hpp [hunk 2]
```diff
@@ -41,7 +41,9 @@
 /// and vice versa. This independency could be important, because, for example, the driver
 /// implements its own cache of verification data and change or GET_RAND() would break it.
 
-static int GET_RAND()
```
Comment (bghimireamd) [inline test/random.hpp:44]: You can ignore this for now. I have create another PR https://github.com/ROCmSoftwarePlatform/MIOpen/pull/1889 that address this issue.

File: include/miopen/miopen.h [hunk 3]
```diff
@@ -2690,6 +2699,27 @@ miopenDestroyActivationDescriptor(miopenActivationDescriptor_t activDesc);
...
+MIOPEN_EXPORT miopenStatus_t miopenCreateGemmDescriptor(miopenGemmDescriptor_t* gemmDesc);
+
+MIOPEN_EXPORT miopenStatus_t miopenInitGemmDescriptor(miopenGemmDescriptor_t gemmDesc,
+                                                      int m_,
+                                                      int n_,
+                                                      int k_,
+                                                      long long int ldA_,
+                                                      long long int ldB_,
+                                                      long long int ldC_,
+                                                      miopenDataType_t dataType_);
```
Comment (pfultz2) [inline include/miopen/miopen.h]: For convolution we dont include the input sizes and types in the convolution descriptor. We should not include it in the gemm descriptor either.

File: src/fusion.cpp [hunk 4]
```diff
@@ -233,6 +233,21 @@ miopenStatus_t ConvForwardOpDescriptor::SetArgs(OperatorArgs& args,
     return miopenStatusSuccess;
 }
 
+miopenStatus_t GemmOpDescriptor::GetOutputDesc(TensorDescriptor& output_desc) const
+{
+    std::vector<int> lens_in = {gemm_descriptor.GetM(), gemm_descriptor.GetN()};
```
Comment (pfultz2) [inline src/fusion.cpp]: The lens should be taken from the tensor descriptors.

File: src/fusion.cpp [hunk 5]
```diff
@@ -233,6 +233,21 @@ miopenStatus_t ConvForwardOpDescriptor::SetArgs(OperatorArgs& args,
...
+    TensorDescriptor temp(input_desc.GetType(), input_desc.GetLayout_t(), lens_in);
```
Comment (pfultz2) [inline src/fusion.cpp]: The permutation(or layout) of the input matrices should not be propagated.

**Comment by junliume** [comment]
@bghimireamd is this PR still WIP?

**Comment by bghimireamd** [comment]
@junliume Yes, should be done next sprint.

**Comment by JehandadKhan** [comment]
Closing this for now since @bghimireamd  is busy with other stuff.

File: src/CMakeLists.txt [hunk 6]
```diff
@@ -82,8 +82,10 @@ endfunction()
 
 set( MIOpen_Source
     activ/problem_description.cpp
+    gemm/problem_description.cpp
     activ_api.cpp
     api/find2_0_commons.cpp
+    gemm_api.cpp
```
Comment (DrizztDoUrden) [inline src/CMakeLists.txt:88]: I'd move this to api/

File: src/CMakeLists.txt [hunk 7]
```diff
@@ -542,7 +546,7 @@ endif()
 if(miopengemm_FOUND OR MIOPEN_USE_ROCBLAS OR MIOPEN_USE_MIOPENTENSILE)
     list(APPEND MIOpen_Source
         gemm_v2.cpp
-        miopengemm.cpp
+        miopengemm.cpp
```
Comment (DrizztDoUrden) [inline src/CMakeLists.txt:549]: Do we return miopengemm?

File: src/include/miopen/gemm/context.hpp [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+namespace miopen {
+struct GemmDesc;
+struct Handle;
+struct TensorDescriptor;
+
+/// A leftover of the legacy design, houses problem config,
+/// environmental context (e.g. HW/SW platform) and solver-specific state.
+///
+/// TODO: These three entities should be made separate.
+struct GemmContext : ExecutionContext
```
Comment (DrizztDoUrden) [inline src/include/miopen/gemm/context.hpp:46]: `ExecutionContext` shouldn't be extended. I have a PR to be merged soon, which among other things moves `is_for_generic_search` to `ExecutionContext`.

File: src/include/miopen/gemm/problem_description.hpp [hunk 9]
```diff
@@ -0,0 +1,123 @@
...
+    friend std::ostream& operator<<(std::ostream& os, const ProblemDescription& obj)
+    {
+        obj.Serialize(os);
+        return os;
+    }
+
+    template <class Self>
+    static void Visit(Self&& self, std::function<void(int, std::string)> f)
+    {
+        // Once we tune gemm kernels we will not ignore.
```
Comment (DrizztDoUrden) [inline src/include/miopen/gemm/problem_description.hpp:101]: This should either actually serialize or not implement `SQLiteSerializable`.
If it is expected, but not implemented yet, it should do `MIOPEN_THROW(miopenStatusNotImplemented)`.

File: src/ocl/fusionopbiasbnactivocl.cpp [hunk 10]
```diff
@@ -47,6 +47,13 @@ miopenStatus_t FusionOpDescriptor::GetNetworkConfig(std::stringstream& /*network
     return miopenStatusSuccess;
 }
 
+miopenStatus_t GemmOpDescriptor::GetNetworkConfig(std::stringstream& network_config,
```
Comment (DrizztDoUrden) [inline src/ocl/fusionopbiasbnactivocl.cpp:50]: This shouldn't take `Handle`.

File: src/solver.cpp [hunk 11]
```diff
@@ -292,6 +292,7 @@ inline SolverRegistrar::SolverRegistrar(IdRegistryData& registry)
              Primitive::Fusion,
              solver::fusion::ConvOclDirectFwdFused{}.SolverDbId(),
              miopenConvolutionAlgoDirect);
+    Register(registry, ++id, Primitive::Fusion, solver::fusion::CKGEMMActiv{}.SolverDbId());
```
Comment (DrizztDoUrden) [inline src/solver.cpp:295]: This should only be added to the end of the file or it would corrupt existing serialized solutions.

**Comment by atamazov** [comment]
@DrizztDoUrden This PR have been closed and branch deleted 2 months ago, any reason to review?

**Comment by bghimireamd** [comment]
@atamazov I am currently tasked with integrating CK's GEMM fusion(GEMM + GELU) solver into MIOpen. This was an old PR of mine that already had done some grunt work, so wanted help from @DrizztDoUrden to see if what I have done so far is correct. We closed this PR last time since there was no proper infrastructure for find fusion solver. Since @DrizztDoUrden now have [find_fusion](https://github.com/ROCmSoftwarePlatform/MIOpen/pull/2388) I will create another PR that will addressing all the issues mentioned here with [find_fusion](https://github.com/ROCmSoftwarePlatform/MIOpen/pull/2388) integrated.

**Comment by atamazov** [comment]
@bghimireamd Thanks for informing and for continuing this work!

**Pull request closed without merging**